File: src/main/java/com/_14ercooper/worldeditor/async/AsyncOperation.java
package com._14ercooper.worldeditor.async;

import java.util.ArrayDeque;
import java.util.List;

import org.bukkit.block.Block;
import org.bukkit.entity.Player;

import com._14ercooper.schematics.SchemLite;
import com._14ercooper.worldeditor.blockiterator.BlockIterator;
import com._14ercooper.worldeditor.blockiterator.iterators.MultiIterator;
import com._14ercooper.worldeditor.operations.Operator;
import com._14ercooper.worldeditor.undo.Undo;

public class AsyncOperation {
    protected String key = "";
    protected Operator operation = null;
    protected ArrayDeque<Block> toOperate = null;
    protected BlockIterator blocks = null;
    protected Player player = null;
    protected Undo undo = null;
    protected boolean undoRunning = false;
    public MultiIterator childOp = null;

    public AsyncOperation(Operator o, Player p, BlockIterator b) {
	key = "iteredit";
	operation = o;
	player = p;
	blocks = b;
    }

    public AsyncOperation(Operator o, BlockIterator b) {
	key = "rawiteredit";
	operation = o;
	blocks = b;
    }

    public AsyncOperation(Operator o) {
	key = "messyedit";
	operation = o;
    }

    // New schematics system
    protected SchemLite schem = null;
    protected int[] origin = {};

    public AsyncOperation(SchemLite sl, boolean saveSchem, int[] o, Player p) {
	schem = sl;
	origin = o;
	blocks = schem.getIterator(origin[0], origin[1], origin[2], p.getWorld());
	if (saveSchem) {
	    key = "saveschem";
	}
	else {
	    key = "loadschem";
	}
	player = p;
    }

    // Selection move/stack
    protected int[] offset = {};
    protected int times = 0;
    protected boolean delOriginal = false;
    // Uses the same iterator as other functions

    public AsyncOperation(BlockIterator selectionIter, int[] cloneOffset, int cloneTimes, boolean delOriginalBlocks,
	    Player p) {
	key = "selclone";
	blocks = selectionIter;
	offset = cloneOffset;
	times = cloneTimes;
	delOriginal = delOriginalBlocks;
	player = p;
    }

    // Multibrush
    List<BlockIterator> iterators;
    List<Operator> operations;

    public AsyncOperation(List<BlockIterator> iterators, List<Operator> operations, Player p) {
	key = "multibrush";
	this.iterators = iterators;
	this.operations = operations;
	player = p;
    }
}

File: src/main/java/com/_14ercooper/worldeditor/async/AsyncOperation.kt
/**
 * This file is part of 14erEdit.
 * 
  * 14erEdit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * 14erEdit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with 14erEdit.  If not, see <https://www.gnu.org/licenses/>.
 */

package com._14ercooper.worldeditor.async

import com._14ercooper.schematics.SchemLite
import com._14ercooper.worldeditor.blockiterator.BlockIterator
import com._14ercooper.worldeditor.operations.Operator
import com._14ercooper.worldeditor.undo.UndoElement
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player

class AsyncOperation {
    val key: String
    var operation: Operator? = null
    var blocks: BlockIterator? = null
    var player: CommandSender
    var undo: UndoElement

    constructor(o: Operator?, p: CommandSender, b: BlockIterator?, thisUndo: UndoElement) {
        key = "iteredit"
        operation = o
        player = p
        blocks = b
        undo = thisUndo
    }

    // New undo system
    var undoList: MutableList<UndoElement>? = null

    constructor(undos: MutableList<UndoElement>, thisPlayer: CommandSender) {
        key = "undoedit"
        player = thisPlayer
        undoList = undos
        undo = undos[0]
    }

    // New schematics system
    var schem: SchemLite? = null
    private var origin = intArrayOf()
    var startedWrite: Boolean = false

    constructor(sl: SchemLite?, saveSchem: Boolean, o: IntArray, p: CommandSender, thisUndo: UndoElement) {
        schem = sl
        origin = o
        blocks = schem!!.getIterator(
            origin[0], origin[1], origin[2], if (p is Player) {
                p.world
            } else {
                Bukkit.getServer().worlds[0]
            }
        )
        key = if (saveSchem) {
            "saveschem"
        } else {
            "loadschem"
        }
        player = p
        undo = thisUndo
    }

    // Selection move/stack
    var offset = intArrayOf()
    var times = 0
    var delOriginal = false

    // Uses the same iterator as other functions
    constructor(
        selectionIter: BlockIterator?, cloneOffset: IntArray, cloneTimes: Int, delOriginalBlocks: Boolean,
        p: CommandSender, thisUndo: UndoElement
    ) {
        key = "selclone"
        blocks = selectionIter
        offset = cloneOffset
        times = cloneTimes
        delOriginal = delOriginalBlocks
        player = p
        undo = thisUndo
    }

    // Multibrush
    lateinit var iterators: MutableList<BlockIterator>
    lateinit var operations: MutableList<Operator>

    constructor(iterators: List<BlockIterator>, operations: List<Operator>, p: CommandSender, thisUndo: UndoElement) {
        key = "multibrush"
        this.iterators = iterators as MutableList<BlockIterator>
        this.operations = operations as MutableList<Operator>
        player = p
        undo = thisUndo
    }
}